lib/network/kmip/transport/TransportTemplate.js

```diff
@@ -0,0 +1,161 @@
+                                this.send(logger,
+                                          deferedRequest.encodedMessage,
+                                          deferedRequest.cb);
+                            });
+                        } else if (this.callbackPipeline.length === 0 &&
+                                   this.deferedRequests.length === 0 &&
+                                   this.pipelineDrainedCallback) {
+                            this.pipelineDrainedCallback();
+                            this.pipelineDrainedCallback = null;
+                        }
+                    });
+                    socket.on('end', () => {
+                        const error = Error('Conversation interrupted');
+                        this._drainQueuesWithError(error);
+                        this.socket = null;
+                    });
+                    socket.on('error', err => {
+                        this._drainQueuesWithError(err);
+                    });
+                    readyCallback(null);
+                });
+            this.socket = socket;
+        } catch (err) {
+            logger.error();
+            readyCallback(err);
+        }
+    }
+
+    _doSend(logger, encodedMessage, cb) {
+        const socket = this.socket;
+        if (!socket) {
+            const error = new Error('Socket to server not available');
+            logger.error('TransportTemplate::_doSend', { error });
+            return cb(error);
+        }
+        this.callbackPipeline.push(cb);
+        socket.cork();
+        socket.write(encodedMessage);
+        socket.uncork();
+        return undefined;
+    }
+
+    /**
+     * Send an encoded message to the server
+     * @param {Object} logger - Werelogs logger object
+     * @param {Buffer} encodedMessage - the encoded message to send to the
+     *                                  server
+     * @param {Function} cb - (err, conversation, rawResponse)
+     * @returns {undefined}
+     */
+    send(logger, encodedMessage, cb) {
+        if (this.callbackPipeline.length >= this.pipelineDepth) {
+            return this.deferedRequests.push({ encodedMessage, cb });
+        }
+        assert(encodedMessage.length !== 0);
+        if (this.socket === null) {
+            return this._createConversation(logger, err => {
+                if (err) {
+                    return cb(err);
+                }
+                return this._doSend(logger, encodedMessage, cb);
+            });
+        }
+        return this._doSend(logger, encodedMessage, cb);
+    }
+
+    /**
+     * Gracefuly interrupt the conversation. If the caller keeps sending
+     * message after calling this function, the conversation won't
+     * converge to its end.
+     * @returns {undefined}
+     */
+    end() {
+        if (this.callbackPipeline.length !== 0 ||
+            this.deferedRequests.length !== 0) {
+            this.pipelineDrainedCallback = this.socket.end.bind(this.socket);
```

Comment:
Is there any scenario where `this.socket` could be `null` here?

Reply:
yes, after a misusage of the API during shutdown (calling twice, or race reaping the pending errors).
Good catch, need to be checked